10-habits/progress/internal/server/server.go

```diff
@@ -55,22 +56,42 @@ func (s *Server) Listen(ctx context.Context, port int) error {
 	grpcServer := s.registerGRPCServer()
 	log.Infof("gRPC server started and listening to port %d", port)
 
+	// Use a channel to report errors from the gRPC server back to
 	errChan := make(chan error)
+	g := errgroup.Group{}
+	defer func() {
+		err := g.Wait()
+		if err != nil {
+			errChan <- fmt.Errorf("error while serving: %w", err)
+		}
+		close(errChan)
+	}()
+
 	// Listen to the port. This will only return when something kills or stops the server.
-	go func() {
+	g.Go(func() error {
 		// This goroutine will be killed when the context is ended at the end of this function.
 		err := grpcServer.Serve(listener)
 		if err != nil {
-			errChan <- fmt.Errorf("error while listening: %w", err)
+			log.Infof("error while serving gRPC: %s", err)
+
+			return fmt.Errorf("gRPC server error: %w", err)
 		}
-	}()
 
-	go func() {
+		return nil
+	})
+
+	g.Go(func() error {
 		const pprofPort = 6060
 		log.Infof("Starting pprof listener on port %d\n", pprofPort)
 		err := http.ListenAndServe(net.JoinHostPort(addr, strconv.Itoa(pprofPort)), nil)
```

Comment:
Nothing is canceling this listener, so some tests should be failing. This indicates that there are tests missing.